packages/network-controller/src/NetworkController.ts

```diff
@@ -896,6 +1156,246 @@ export class NetworkController extends BaseControllerV2<
+      const networkClientId = buildInfuraNetworkClientId(network);
+      const networkClientConfiguration: InfuraNetworkClientConfiguration = {
+        type: NetworkClientType.Infura,
+        network,
+        infuraProjectId: this.#infuraProjectId,
+      };
+      return [
+        NetworkClientType.Infura,
+        networkClientId,
+        networkClientConfiguration,
+      ];
+    });
+  }
+
+  /**
+   * Constructs the list of network clients for custom networks (that is, those
+   * which consumers have added via `networkConfigurations`).
+   *
+   * @returns The network clients.
+   */
+  #buildIdentifiedCustomNetworkClientConfigurations(): [
+    NetworkClientType.Custom,
+    CustomNetworkClientId,
+    CustomNetworkClientConfiguration,
+  ][] {
+    return Object.entries(this.state.networkConfigurations).map(
+      ([networkConfigurationId, networkConfiguration]) => {
+        if (networkConfiguration.chainId === undefined) {
+          throw new Error('chainId must be provided for custom RPC endpoints');
+        }
+        if (networkConfiguration.rpcUrl === undefined) {
+          throw new Error('rpcUrl must be provided for custom RPC endpoints');
+        }
+        const networkClientId = buildCustomNetworkClientId(
+          networkConfigurationId,
+        );
+        const networkClientConfiguration: CustomNetworkClientConfiguration = {
+          type: NetworkClientType.Custom,
+          chainId: networkConfiguration.chainId,
+          rpcUrl: networkConfiguration.rpcUrl,
+        };
+        return [
+          NetworkClientType.Custom,
+          networkClientId,
+          networkClientConfiguration,
+        ];
+      },
+    );
+  }
+
+  /**
+   * Converts the provider config object in state to a network client
+   * configuration object.
+   *
+   * @returns The network client config.
+   * @throws If the provider config is of type "rpc" and lacks either a
+   * `chainId` or an `rpcUrl`.
+   */
+  #buildIdentifiedNetworkClientConfigurationFromProviderConfig():
+    | [
+        NetworkClientType.Custom,
+        CustomNetworkClientId,
+        CustomNetworkClientConfiguration,
+      ]
+    | [
+        NetworkClientType.Infura,
+        BuiltInNetworkClientId,
+        InfuraNetworkClientConfiguration,
+      ] {
+    const { providerConfig } = this.state;
+
+    if (isCustomProviderConfig(providerConfig)) {
+      validateCustomProviderConfig(providerConfig);
+      const networkClientId = buildCustomNetworkClientId(
+        providerConfig,
+        this.state.networkConfigurations,
+      );
+      const networkClientConfiguration: CustomNetworkClientConfiguration = {
+        chainId: providerConfig.chainId,
+        rpcUrl: providerConfig.rpcUrl,
+        type: NetworkClientType.Custom,
+      };
+      return [
+        NetworkClientType.Custom,
+        networkClientId,
+        networkClientConfiguration,
+      ];
+    }
+
+    if (isInfuraProviderConfig(providerConfig)) {
```

Comment:
Nit: Maybe I'm missing something, but.... could we skip building the client in this case? It seems guaranteed that it'd already exist (all of the built-in network clients are always built).

Reply:
You're right! Updated in [d6d1dbd](https://github.com/MetaMask/core/pull/1439/commits/d6d1dbd2ede0990c7e054e626058bc1f0b2efcb2).